rust/core-lib/src/view.rs

```diff
@@ -125,8 +126,48 @@ impl View {
         self.cursor_col = col;
     }
 
-    pub fn get_cursor_col(&mut self) -> usize {
-        self.cursor_col
+    /// Toggles a caret at the given offset.
+    pub fn toggle_sel(&mut self, offset: usize) {
+        self.sel_dirty = true;
+        if !self.selection.regions_in_range(offset, offset).is_empty() {
+            self.selection.delete_range(offset, offset);
+            if !self.selection.is_empty() {
+                return;
+            }
+        }
+        let region = SelRegion {
+            start: offset,
+            end: offset,
+            // TODO: might want to set horiz to some meaningful value.
+            horiz: None,
```

Comment:
I'm rethinking HorizPos compared with the old cursor_col now. The latter was basically always calculated eagerly. Most operations that set the cursor, other than vertical movement, would set it. My new thinking (which is associated with the fact that it's now an `Option`) is that these operations would reset it, and the first vertical motion would transition it from None->Some.

Computing HorizPos, if we were going to stick with that, does require view and text, but those are not too hard to plumb.

I have another patch on my disk that has most movements implemented. Not sure whether it's best to push it to this branch or do it in another PR. It is, however, still based on computing HorizPos eagerly. I think I want to think on this a little more.